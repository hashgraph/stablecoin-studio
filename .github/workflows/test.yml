name: test
on:
  pull_request:
    branches: [ main, release/**, feature/** ]
  push:
    branches: [ main, release/** ]
    tags: [ v* ]
jobs:
  test-node:
    name: Unit Tests
    runs-on: ubuntu-latest
    steps:
      - name: Use Node.js TLS 16.x
        uses: actions/setup-node@v1
        with:
          node-version: 16.x
      - name: Checkout repository
        uses: actions/checkout@v2
      - name: Install dependencies
        run: npm i
      - name: Build
        run: npm run build
      - name: Test
        run: npm run test