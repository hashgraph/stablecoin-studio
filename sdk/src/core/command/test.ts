/* eslint-disable @typescript-eslint/no-empty-interface */

interface CommandResponse {}
interface ICommand {}

export class Command<T> implements ICommand {
	private $resultType!: T;
}

export interface ICommandBus {
	execute<X>(command: Command<X>): Promise<X>;
}

export type IInferringCommandHandler<CommandType extends Command<unknown>> =
	CommandType extends Command<infer ResultType>
		? {
				execute(command: CommandType): Promise<ResultType>;
		  }
		: never;

class ConcreteCommandResponse implements CommandResponse {
	constructor(public readonly status: string) {}
}

class ConcreteCommand extends Command<ConcreteCommandResponse> {
	constructor(public readonly value: string) {
		super();
	}
}

class CommandHandler<CommandType extends Command<unknown>>
	implements IInferringCommandHandler<ConcreteCommand>
{
	execute(command: ConcreteCommand): Promise<ConcreteCommandResponse> {
		throw new Error('Method not implemented.');
	}
}

class CommandBus implements ICommandBus {
	execute<X>(command: Command<X>): Promise<X> {
		throw new Error('Method not implemented.');
	}
}

// Test

const bus = new CommandBus();
const res = bus.execute(new ConcreteCommand('1'));
