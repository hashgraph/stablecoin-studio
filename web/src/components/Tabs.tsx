import {
	Tabs as ChakraTabs,
	TabsProps as ChakraTabsProps,
	TabList,
	TabPanels,
	Tab,
	TabPanel,
	Text,
	Box,
	Divider,
	Flex,
	Button,
	VStack,
} from '@chakra-ui/react';
import { MouseEvent, ReactNode, useState } from 'react';

export interface OptionBaseTabProps {
	number: string;
	title: string;
	complete: boolean;
	children: ReactNode;
}

interface BaseTabsProps extends Omit<ChakraTabsProps, 'children'> {
	options: OptionBaseTabProps[];
	textDefaultButtonPrimary?: string;
	textLastButtonPrimary?: string;
	textDefaultButtonSecondary?: string;
	textFirstButtonSecondary?: string;
	handleFirstButtonSecondary: () => void;
	handleLastButtonPrimary: () => void;
}

const Tabs = (props: BaseTabsProps) => {
	const {
		options,
		textDefaultButtonPrimary,
		textLastButtonPrimary,
		textDefaultButtonSecondary,
		textFirstButtonSecondary,
		handleFirstButtonSecondary,
		handleLastButtonPrimary,
	} = props;

	const [tabIndex, setTabIndex] = useState(0);

	const handleStep = (e: MouseEvent<HTMLButtonElement>, index: number, type: string) => {
		e.preventDefault();

		return type === 'next' ? setTabIndex(index + 1) : setTabIndex(index - 1);
	};

	return (
		<ChakraTabs
			isFitted
			isLazy
			variant='simple'
			index={tabIndex}
			onChange={(index) => setTabIndex(index)}
			position='relative'
			h='full'
		>
			<TabList justifyContent='space-between'>
				{options.map((option, index) => {
					const { title, number } = option;

					const currentTab = tabIndex === index;

					return (
						<Tab key={index} p='15px'>
							<Text
								data-testid={`tabs-tab-number-${index + 1}`}
								fontSize='14px'
								color={currentTab ? 'brand.primary' : 'brand.gray600'}
								mr='10px'
							>
								{number}
							</Text>
							<Text
								data-testid={`tabs-tab-title-${index + 1}`}
								fontSize='14px'
								color='brand.gray600'
							>
								{title}
							</Text>
						</Tab>
					);
				})}
			</TabList>

			<TabPanels h='calc(100% - 55px)' bgColor='brand.gray100'>
				{options.map((option, index) => {
					const { children } = option;

					const firstTab = index === 0;
					const lastTab = index === options.length - 1;

					return (
						<TabPanel
							data-testid={`tabs-tabpanel-container-${index + 1}`}
							h='full'
							p='0px'
							key={index}
						>
							<VStack h='full' spacing={4} align='stretch'>
								<Box h='calc(100% - 85px)' overflowY='auto'>
									{children}
								</Box>
								<Box h='88px'>
									<Divider color='brand.gray2' />
									<Flex p='0px 24px' justifyContent='flex-end' h='99%' alignItems='center'>
										<Button
											data-testid={`tabs-tabpanel-button-secondary-${index + 1}`}
											variant='secondary'
											mr='20px'
											onClick={
												firstTab ? handleFirstButtonSecondary : (e) => handleStep(e, index, 'prev')
											}
										>
											{firstTab
												? textFirstButtonSecondary || 'Cancel'
												: textDefaultButtonSecondary || 'Go back'}
										</Button>
										<Button
											data-testid={`tabs-tabpanel-button-primary-${index + 1}`}
											variant='primary'
											onClick={
												lastTab ? handleLastButtonPrimary : (e) => handleStep(e, index, 'next')
											}
										>
											{lastTab
												? textLastButtonPrimary || 'Finish'
												: textDefaultButtonPrimary || 'Next step'}
										</Button>
									</Flex>
								</Box>
							</VStack>
						</TabPanel>
					);
				})}
			</TabPanels>
		</ChakraTabs>
	);
};

export default Tabs;
