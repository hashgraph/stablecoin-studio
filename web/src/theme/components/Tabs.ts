export const Tabs = {
	variants: {
		simple: {
			tab: {
				borderBottom: '1px solid',
				borderBottomColor: 'brand.gray300',
				_selected: {
					borderBottom: '4px solid',
					borderBottomColor: 'light.purple2',
					_focus: {
						boxShadow: 'none',
					},
				},
			},
		},
	},
};
