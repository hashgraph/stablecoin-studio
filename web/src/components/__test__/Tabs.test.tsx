import { Box } from '@chakra-ui/react';
import { fireEvent, act } from '@testing-library/react';
import { render } from '../../test/index';
import Tabs, { OptionBaseTabProps } from '../Tabs';

const options: OptionBaseTabProps[] = [
	{
		number: '01',
		title: 'Option 1',
		complete: false,
		children: <Box>Tab 1</Box>,
	},
	{
		number: '02',
		title: 'Option 2',
		complete: false,
		children: <Box>Tab 2</Box>,
	},
];

const tabsProps = {
	options,
	textLastButtonPrimary: 'Last button',
	handleLastButtonPrimary: () => jest.fn,
	handleFirstButtonSecondary: () => jest.fn,
};

describe(`<${Tabs.name} />`, () => {
	test('should render correctly', () => {
		const component = render(<Tabs {...tabsProps} />);

		expect(component.asFragment()).toMatchSnapshot();
	});

	test('should have number and title each tabs', () => {
		const component = render(<Tabs {...tabsProps} />);

		const titleTab1 = component.getByTestId('tabs-tab-title-1');
		const titleTab2 = component.getByTestId('tabs-tab-title-2');

		const numberTab1 = component.getByTestId('tabs-tab-number-1');
		const numberTab2 = component.getByTestId('tabs-tab-number-2');

		expect(titleTab1).toHaveTextContent(options[0].title);
		expect(titleTab2).toHaveTextContent(options[1].title);
		expect(numberTab1).toHaveTextContent(options[0].number);
		expect(numberTab2).toHaveTextContent(options[1].number);
	});

	test('should change tab when next step button clicked', async () => {
		const component = render(<Tabs {...tabsProps} />);

		const buttonPrimaryTab1 = component.getByTestId('tabs-tabpanel-button-primary-1');

		act(() => {
			fireEvent.click(buttonPrimaryTab1);
		});

		await act(async () => {
			const buttonSecondaryTab2 = component.getByTestId('tabs-tabpanel-button-secondary-2');

			expect(buttonSecondaryTab2).toBeInTheDocument();
		});
	});

	test('Cancel button should call function passed by handleFirstButtonSecondary prop', async () => {
		const handleFirstButtonSecondary = jest.fn();

		const component = render(
			<Tabs {...tabsProps} handleFirstButtonSecondary={handleFirstButtonSecondary} />,
		);

		const buttonSecondaryTab1 = component.getByTestId('tabs-tabpanel-button-secondary-1');

		act(() => {
			fireEvent.click(buttonSecondaryTab1);
		});

		await act(async () => {
			expect(handleFirstButtonSecondary).toHaveBeenCalled();
		});
	});

	test('Last button should call function passed by handleLastButtonPrimary prop', async () => {
		const handleLastButtonPrimary = jest.fn();

		const component = render(
			<Tabs {...tabsProps} handleLastButtonPrimary={handleLastButtonPrimary} />,
		);

		const buttonPrimaryTab1 = component.getByTestId('tabs-tabpanel-button-primary-1');

		act(() => {
			fireEvent.click(buttonPrimaryTab1);
		});

		await act(async () => {
			const buttonPrimaryTab2 = component.getByTestId('tabs-tabpanel-button-primary-2');

			act(() => {
				fireEvent.click(buttonPrimaryTab2);
			});

			expect(handleLastButtonPrimary).toHaveBeenCalled();
		});
	});
});
