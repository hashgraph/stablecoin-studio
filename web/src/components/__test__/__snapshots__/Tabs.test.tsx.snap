// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-tabs css-1x6vjw3"
  >
    <div
      aria-orientation="horizontal"
      class="chakra-tabs__tablist css-10a73ai"
      role="tablist"
    >
      <button
        aria-controls="tabs-:r0:--tabpanel-0"
        aria-selected="true"
        class="chakra-tabs__tab css-1abs3h5"
        data-index="0"
        id="tabs-:r0:--tab-0"
        role="tab"
        tabindex="0"
        type="button"
      >
        <p
          class="chakra-text css-bak5wb"
          data-testid="tabs-tab-number-1"
        >
          01
        </p>
        <p
          class="chakra-text css-57h6bf"
          data-testid="tabs-tab-title-1"
        >
          Option 1
        </p>
      </button>
      <button
        aria-controls="tabs-:r0:--tabpanel-1"
        aria-selected="false"
        class="chakra-tabs__tab css-1abs3h5"
        data-index="1"
        id="tabs-:r0:--tab-1"
        role="tab"
        tabindex="-1"
        type="button"
      >
        <p
          class="chakra-text css-1bxd6jy"
          data-testid="tabs-tab-number-2"
        >
          02
        </p>
        <p
          class="chakra-text css-57h6bf"
          data-testid="tabs-tab-title-2"
        >
          Option 2
        </p>
      </button>
    </div>
    <div
      class="chakra-tabs__tab-panels css-27a8z"
    >
      <div
        aria-labelledby="tabs-:r0:--tab-0"
        class="chakra-tabs__tab-panel css-4278sh"
        data-testid="tabs-tabpanel-container-1"
        id="tabs-:r0:--tabpanel-0"
        role="tabpanel"
        tabindex="0"
      >
        <div
          class="chakra-stack css-1eot3nj"
        >
          <div
            class="css-1l1zmw3"
          >
            <div
              class="css-0"
            >
              Tab 1
            </div>
          </div>
          <div
            class="css-1q6t3f6"
          >
            <hr
              aria-orientation="horizontal"
              class="chakra-divider css-12thghs"
            />
            <div
              class="css-g4d8ui"
            >
              <button
                class="chakra-button css-124znom"
                data-testid="tabs-tabpanel-button-secondary-1"
                type="button"
              >
                Cancel
              </button>
              <button
                class="chakra-button css-1jvauvu"
                data-testid="tabs-tabpanel-button-primary-1"
                type="button"
              >
                Next step
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        aria-labelledby="tabs-:r0:--tab-1"
        class="chakra-tabs__tab-panel css-4278sh"
        data-testid="tabs-tabpanel-container-2"
        hidden=""
        id="tabs-:r0:--tabpanel-1"
        role="tabpanel"
        tabindex="0"
      />
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
